functions/slack/SlackKnowledgeGraphSearch/Function.cs

```diff
@@ -0,0 +1,175 @@
+using Google.Apis.Kgsearch.v1.Data;
+using Google.Apis.Services;
+using Google.Cloud.Functions.Framework;
+using Google.Cloud.Functions.Invoker;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+[assembly: FunctionsStartup(typeof(SlackKnowledgeGraphSearch.Startup))]
+
+namespace SlackKnowledgeGraphSearch
+{
+    // [START functions_slack_setup]
+    public class Startup : FunctionsStartup
+    {
+        private static readonly TimeSpan SlackTimestampTolerance = TimeSpan.FromMinutes(5);
+
+        public override void ConfigureServices(WebHostBuilderContext context, IServiceCollection services)
+        {
+            // These can come from an environment variable or a configuration file.
+            string kgApiKey = context.Configuration["KG_API_KEY"];
+            string slackSigningSecret = context.Configuration["SLACK_SECRET"];
+
+            services.AddSingleton(new KgsearchService(new BaseClientService.Initializer
+            {
+                ApiKey = kgApiKey
+            }));
+            services.AddSingleton(new SlackRequestVerifier(slackSigningSecret, SlackTimestampTolerance));
+            base.ConfigureServices(context, services);
+        }
+    }
+    // [END functions_slack_setup]
+
+    public class Function : IHttpFunction
+    {
+        private readonly ILogger _logger;
+        private readonly KgsearchService _kgService;
+        private readonly SlackRequestVerifier _verifier;
+
+        public Function(ILogger<Function> logger, KgsearchService kgService, SlackRequestVerifier verifier) =>
+            (_logger, _kgService, _verifier) = (logger, kgService, verifier);
+
+        // [START functions_slack_search]
+        public async Task HandleAsync(HttpContext context)
+        {
+            var request = context.Request;
+            var response = context.Response;
+
+            // Validate request
+            if (request.Method != "POST")
+            {
+                _logger.LogWarning("Unexpected request method '{method}'", request.Method);
+                response.StatusCode = (int) HttpStatusCode.MethodNotAllowed;
+                return;
+            }
+
+            if (!request.HasFormContentType)
+            {
+                _logger.LogWarning("Unexpected content type '{contentType}'", request.ContentType);
+                response.StatusCode = (int) HttpStatusCode.BadRequest;
+                return;
+            }
+
+            // We need to read the request body twice: once to validate the signature,
+            // and once to read the form content. We copy it into a memory stream,
+            // so that we can rewind it after reading.
+            var bodyCopy = new MemoryStream();
+            await request.Body.CopyToAsync(bodyCopy);
```

Comment:
ConfigureAwait missing here, but it's called a few lines down, consistency might be best.

Reply:
Will remove it later on - it's unnecessary in ASP.NET Core.